tracer/src/Datadog.Trace/Span.cs

```diff
@@ -37,12 +38,17 @@ internal Span(SpanContext context, DateTimeOffset? start)
         {
         }
 
-        internal Span(SpanContext context, DateTimeOffset? start, ITags tags)
+        internal Span(SpanContext context, DateTimeOffset? start, ITags tags, IEnumerable<SpanLink> links = null)
         {
             Tags = tags ?? new CommonTags();
             Context = context;
             StartTime = start ?? Context.TraceContext.Clock.UtcNow;
 
+            foreach (var link in links ?? Enumerable.Empty<SpanLink>())
+            {
+                AddLink(link);
+            }
```

Comment:
🚓 heap allocation police 🚓
Although using `??` is more concise, if we check for `links == null` first we can skip the iterator allocation.
```suggestion
            if (links is not null)
            {
                foreach (var link in links)
                {
                    AddLink(link);
                }
            }
```

Reply:
Also, if you can change `IEnumerable<T>` to `List<T>`, that iterator is a `struct`, so no heap allocation even when `link` is _not_ null.

Reply:
I've added the early null check as suggested 😄 

For the change to `Lis<T>`, I technically _can_ change from `IEnumerable<T>` to `List<T>` everywhere, but API-wise it seems better to be using `IEnumerable<T>`. In the common case, I expect this to be null so I think performance should be acceptable